docs/ERROR_HANDLING.md

```diff
@@ -0,0 +1,86 @@
+# Error handling for various entry points
+
+In this document we discuss how different types of errors during contract
+execution are handled by wasmd and the blockchain.
+
+## Two levels of errors
+
+When cosmwasm-vm executes a contract, the caller receives a nested result type:
+`VmResult<ContractResult<R>>` with some success response `R`. The outer
+`VmResult` is created by the host environment and the inner `ContractResult` is
+created inside of the contract. Most application specific error should go into
+`ContractResult` errors. This is what happens when you use `?` inside of your
+contract implementations. The `VmResult`
+[error cases](https://github.com/CosmWasm/cosmwasm/blob/v1.2.3/packages/vm/src/errors/vm_error.rs#L11-L148)
+include e.g.
+
+- Caching errors such as a missing Wasm file or corrupted module
+- Serialization problems in the contract-host communication
+- Panics from panic handler in contract
+- Errors in crypto API calls
+- Out of gas
+- Unreachable statements in the Wasm bytecode
+
+## Error handling
+
+In wasmvm those two error types are merged into one and handled as one thing in
+the caller (wasmd):
+
+- [Instantiate](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L144-L151)
+- [Execute](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L192-L199)
+- [Migtate](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L275-L282)
+- [Sudo](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L318-L325)
+- [Reply](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L363-L370)
+- [IBCChannelOpen](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L406-L413)
+- [IBCChannelConnect](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L449-L456)
+- [IBCChannelClose](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L492-L499)
+- [IBCPacketAck](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L576-L583)
+- [IBCPacketTimeout](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L620-L627)
+
+However, there is one exception:
+
+- [IBCPacketReceive](https://github.com/CosmWasm/wasmvm/blob/v1.2.0/lib.go#L535-L539)
+
+Instead of returning only the contents of the `Ok` case, the whole
+`IBCReceiveResult` is returned. This allows the caller to handle the two layers
+of errors differently.
+
+As pointed out by our auditors from Oak Security, this
+[is inconsistent](https://github.com/CosmWasm/wasmvm/issues/398). Historically
+merging the two error types was the desired behaviour. When `IBCPacketReceive`
+came in, we needed the differentiation to be available in wasmd, which is why
+the API is different than the others. Ideally we always return the contract
+Result and let wasmd handle it.
+
+## Handing ibc_packet_receive errors
+
+For wasmd before 0.32, contract errors and VM errors were handled the same. They
```

Comment:
This is very clear (and much better than my text here), but it misses where the original idea came from. Maybe capture this workflow (or is that TMI and we just focus on the 0.32 change, as anything before 0.45 was ancient and no one did IBC then anyway)?

Until 0.X it was handled like the rest where any error reverted the whole transaction.
In order to improve developer ergonomics, in 0.X (the 0.45 upgrade release) we changed this such that both cases revert state change but return error acknoweldgement packet.
In 0.32, we fully implemented the discussion here (link to issue Alex posted) and differentiate between these two cases.

Reply:
I'd indeed focus on just the two stages:
- From [0.22.0](https://github.com/CosmWasm/wasmd/tree/v0.22.0) (2022-01-20, with https://github.com/CosmWasm/wasmd/pull/717) to 0.31
- From 0.32 and beyond

But I can try to expand a bit why they have a different behaviour without going back all the way to pre-0.22 behavior.

Reply:
Done in 6f8b4ec41050b588c98354d4f806e2b5b64a8945